### PR TITLE
New Bid Adapter: BidFabrik

### PR DESCRIPTION
**Pull request by v0idxyz**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [ ] New bidder adapter  
- [ ] Updated bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes

- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

Maintainer: bern@revantage.io
Test Params:
{
  bidder: '<bidder name>',
  params: {
    feed:"oubjsrii",
    host:"bid.bidfabrik.com"
  }
}

### CODE REVIEW

File: modules/bidfabrikBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
+import { registerBidder } from '../src/adapters/bidderFactory.js';
+import { ortbConverter } from '../libraries/ortbConverter/converter.js';
+import { BANNER, VIDEO, NATIVE } from '../src/mediaTypes.js';
+import { deepSetValue, replaceAuctionPrice, triggerPixel, isStr, isArray, logError } from '../src/utils.js';
+
+const BIDDER_CODE = 'bidfabrik';
+const GVLID = 1588; // TODO: confirm this is the GVL ID registered for BidFabrik
```
Comment (robertrmartinez) [inline modules/bidfabrikBidAdapter.js]: please confirm the registered GVL id for BidFabrik and drop the TODO before this merges.

File: modules/revantageBidAdapter.js [hunk 2]
```diff
@@ -1,408 +0,0 @@
-import { registerBidder } from '../src/adapters/bidderFactory.js';
```
Comment (robertrmartinez) [inline modules/revantageBidAdapter.js:1]: Hi @v0idxyz

I do not think we can remove an adapter on master mid-major version like this... we can delete in Prebid 12 see: https://github.com/prebid/Prebid.js/blob/master/PREBID_VERSIONING_DEPRECATION.md

here is an option we can support see this PR for example: https://github.com/prebid/Prebid.js/pull/15022

I have started a discussion with the prebid dev team to determine our official stance will get back to you.

File: modules/bidfabrikBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  const host = params && isStr(params.host) ? params.host.trim() : '';
+  return host || DEFAULT_HOST;
+}
+
+export const spec = {
+  code: BIDDER_CODE,
+  gvlid: GVLID,
+  aliases: [
+    { code: 'revbid', gvlid: GVLID },
+    { code: 'revantage', gvlid: GVLID },
```
Comment (robertrmartinez) [inline modules/bidfabrikBidAdapter.js]: the old revantage adapter validated on `feedId` / `placementId` / `publisherId`. now revantage aliases onto bidfabrik which requires `params.feed`.

So while its not a big deal, if a pub is thinking a smooth transition no work it will not in fact work. But that is a decision your company should make.

File: modules/bidfabrikBidAdapter.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    if (!response || !response.body) {
+      return [];
+    }
+    return converter.fromORTB({
+      response: response.body,
+      request: request.data,
+    }).bids;
+  },
+
+  getUserSyncs(syncOptions, serverResponses) {
```
Comment (robertrmartinez) [inline modules/bidfabrikBidAdapter.js]: the old `getUserSyncs` appended `gdpr` / `gdpr_consent` / `us_privacy` / `gpp` / `gpp_sid` to the sync urls, this one fires whatever the server returns as-is.

thats fine if the LB is baking consent into the urls it returns (it has the request consent), but if not we are firing sync pixels with no consent signal.

if the server handles it can you add a comment saying so, otherwise we prob need to append the consent params here.

File: test/spec/modules/bidfabrikBidAdapter_spec.js [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+      expect(triggerStub.calledOnce).to.equal(true);
+      expect(triggerStub.firstCall.args[0]).to.equal('https://bid.bidfabrik.com/win?p=1.5');
+    });
+
+    it('does nothing without an nurl', function () {
+      spec.onBidWon({ cpm: 1.5 });
+      expect(triggerStub.called).to.equal(false);
+    });
+  });
+});
```
Comment (robertrmartinez) [inline test/spec/modules/bidfabrikBidAdapter_spec.js:457]: `supportedMediaTypes` lists video and native and the converter does handle those for you, but theres no test proving it for this adapter.

please add unit tests to cover this also! at minimum:

- `interpretResponse` for video (`mtype:2`) and native (`mtype:4`) so we know the declared media types actually work
- consent (gdpr / usp / gpp) actually landing in the ortb request body
- multiple seatbids / multiple bids, and `price` 0 / missing `adm`
- `getUserSyncs` consent behavior

**Review by robertrmartinez — CHANGES_REQUESTED** [review]
Several changes + a follow up I will figure out

Also there is no `bidfabrikBidAdapter.md` and no docs PR linked.

for a new bidder please link the prebid.github.io docs PR so the params (feed, host) are documented.

File: modules/bidfabrikBidAdapter.js (hunk 1 above)
Comment (v0idxyz) [reply modules/bidfabrikBidAdapter.js]: Hi!
BidFabrik is a Whitelabel Adapter and will not operate under it's own BidFabrik Name but rather the aliases that are linked to it, that's why I didn't think we'd need the .md
GVL ID is registered to ReVantage GmbH, which operates the WL Adserver

File: modules/revantageBidAdapter.js (hunk 2 above)
Comment (v0idxyz) [reply modules/revantageBidAdapter.js:1]: I merged this one on accident. This PR had nothing to do with revantage, that's why I removed the Code again!

File: modules/bidfabrikBidAdapter.js (hunk 3 above)
Comment (v0idxyz) [reply modules/bidfabrikBidAdapter.js]: Transitions will work by switching out the Adapters completely

File: modules/bidfabrikBidAdapter.js (hunk 4 above)
Comment (v0idxyz) [reply modules/bidfabrikBidAdapter.js]: added that into bf now too

File: test/spec/modules/bidfabrikBidAdapter_spec.js (hunk 5 above)
Comment (v0idxyz) [reply test/spec/modules/bidfabrikBidAdapter_spec.js:457]: added that too

File: modules/bidfabrikBidAdapter.js [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+export const spec = {
+  code: BIDDER_CODE,
+  gvlid: GVLID,
+  aliases: [
+    { code: 'revbid', gvlid: GVLID },
+    { code: 'revantage', gvlid: GVLID },
+  ],
+  supportedMediaTypes: [BANNER, VIDEO, NATIVE],
```
[the remark this answers is not in the document]
Comment (v0idxyz) [reply modules/bidfabrikBidAdapter.js:81]: restored revantage, rename adapter to revantagex

**Comment by robertrmartinez** [comment]
you restored the revantage adapter but not its MD or SPEC files! @v0idxyz

File: modules/bidfabrikBidAdapter.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (patmmccann) [reply modules/bidfabrikBidAdapter.js:81]: That would cause conflicts with keys set in gam, first six characters must be unique

[the remark this answers is not in the document]
Comment (v0idxyz) [reply modules/bidfabrikBidAdapter.js:81]: > That would cause conflicts with keys set in gam, first six characters must be unique

didnt think of gam, thanks for the heads up, moving the x to the front then

File: modules/bidfabrikBidAdapter.js [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+});
+
+function resolveHost(params) {
+  // Publishers may paste a host with a scheme, path, query, or fragment
+  // (e.g. "https://us.bidfabrik.com/foo"); keep only the authority so the
+  // request URL can't end up double-schemed or pointing at an unexpected
+  // path. Fall back to the default if what's left is empty or has whitespace.
+  const raw = params && isStr(params.host) ? params.host.trim() : '';
+  const host = raw.replace(/^https?:\/\//i, '').split('/')[0].split('?')[0].split('#')[0];
+  return host && !/\s/.test(host) ? host : DEFAULT_HOST;
```
[the remark this answers is not in the document]
Comment (v0idxyz) [reply modules/bidfabrikBidAdapter.js]: you told me earlier what to put in there

File: metadata/modules.json [hunk 8]
```diff
@@ -6673,6 +6673,27 @@
       "componentType": "analytics",
       "componentName": "zeta_global_ssp",
       "gvlid": null
+    },
```
Comment (patmmccann) [inline metadata/modules.json:6605]: you do not need to edit this file if you have a gvlid

File: modules/bidfabrikBidAdapter.js [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+  code: BIDDER_CODE,
+  gvlid: GVLID,
+  aliases: [
+    { code: 'revortb', gvlid: GVLID },
+    { code: 'xrevantage', gvlid: GVLID },
+  ],
+  supportedMediaTypes: [BANNER, VIDEO, NATIVE],
+
+  isBidRequestValid(bid) {
+    const feed = bid && bid.params && bid.params.feed;
```
Comment (patmmccann) [inline modules/bidfabrikBidAdapter.js:84]: make sure to type your params with a d.ts file and import it or convert to ts

**Review by patmmccann — CHANGES_REQUESTED** [review]
